### PR TITLE
JACK PipeWire fixes (escape regex chars)

### PR DESCRIPTION
**Pull request by Be-ing**
This fixes various bugs in the JACK host API exposed by PipeWire. Refer to the commit messages for details. I think this should be included in the upcoming 19.7 release because Linux distros are about to switch to PipeWire by default in the next few months (Fedora 34 and Ubuntu 21.04 at least).

I usually write C++ and don't have much experience with C so please point out if I'm doing something wrong. The `replace_string` function might not be the most efficient, but it works, and this is not performance critical code or we wouldn't be using regexes here already.

### CODE REVIEW

File: src/hostapi/jack/pa_jack.c
Comment (daschuer) [inline src/hostapi/jack/pa_jack.c:458]: The name of the variable is misleading, because sizeof normally includes the null terminator, len not.
```suggestion
        size_t original_len = strlen(string);
```

File: src/hostapi/jack/pa_jack.c
Comment (daschuer) [inline src/hostapi/jack/pa_jack.c:471]: This is unsave, because you potentially write behind the original string without permission.
You should take a buffer size parameter as well to be save.

File: src/hostapi/jack/pa_jack.c
Comment (daschuer) [inline src/hostapi/jack/pa_jack.c:463]: You can get around this malloc and free, by copy the tailing part byte by byte to the new location. memmove() is your friend here.

**Comment by Be-ing** [comment]
Regarding the colons in device names, I have [opened an issue on PipeWire's tracker](https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/714) because this violates an old JACK convention. Carla also has problems with `XONE:K2`

File: src/hostapi/jack/pa_jack.c
Comment (daschuer) [inline src/hostapi/jack/pa_jack.c:479]: This should also take a buffer size parameter.

File: src/hostapi/jack/pa_jack.c
Comment (daschuer) [inline src/hostapi/jack/pa_jack.c:639]: strcpy is not save, because it does no length check. A one stop shopping function from the standard library is snprintf() with "%s".

**Review by daschuer — CHANGES_REQUESTED** [review]

File: src/hostapi/jack/pa_jack.c
Comment (daschuer) [inline src/hostapi/jack/pa_jack.c:454]:
```suggestion
void replace_string(char *string, size_t size, const char *to_find, const char *replacement)
```

**Comment by Be-ing** [comment]
> Regarding the colons in device names, I have opened an issue on PipeWire's tracker because this violates an old JACK convention. Carla and QJackCtl also have problems with XONE:K2

This was fixed in PipeWire.

File: src/hostapi/jack/pa_jack.c
Comment (Be-ing) [reply src/hostapi/jack/pa_jack.c]: It now checks the buffer size before memmove. Is this an acceptable alternative?

**Comment by Be-ing** [comment]
Ugh I have been testing the wrong library because the CMake implementation installs to the wrong directory (`lib` instead of `lib64`). The last few commits are broken, working on it...

**Comment by Be-ing** [comment]
Okay I think the code is safe now. I cleaned up the Git history and rebased. Please review again.

**Comment by Be-ing** [comment]
With the current PipeWire master branch, this is working well with all my audio interfaces.

File: src/hostapi/jack/pa_jack.c
Comment (daschuer) [inline src/hostapi/jack/pa_jack.c:455]: buffer_length is ambiguous here, "buffer" is unrelated to \0 while the therm length usually excludes it.  Without const, we cannot use string literals. 
```suggestion
void replace_string(char *string, size_t string_size, const char *to_find, const char *replacement)
```

File: src/hostapi/jack/pa_jack.c
Comment (daschuer) [inline src/hostapi/jack/pa_jack.c:470]: This looks wrong, because string is a negative pointer. You add string below anyway in almost all cases.

File: src/hostapi/jack/pa_jack.c
Comment (daschuer) [inline src/hostapi/jack/pa_jack.c:473]: I think the copied num is wrong. We can also move the null terminator here. How about this. 
memmove(found_position + replacement_length,
                found_position + found_length,
                string + original_length - (found_position + found_length) + 1);

File: src/hostapi/jack/pa_jack.c
Comment (daschuer) [inline src/hostapi/jack/pa_jack.c:474]: simplify to: 
```suggestion
        memcpy(found_position, replacement, replacement_length);
```

File: src/hostapi/jack/pa_jack.c
Comment (daschuer) [inline src/hostapi/jack/pa_jack.c:475]: This is redundant if we copy the null terminator above.

File: src/hostapi/jack/pa_jack.c
Comment (daschuer) [inline src/hostapi/jack/pa_jack.c:477]:
```suggestion
        found_position = strstr(found_position + replacement_length, to_find);
```

**Review by daschuer — CHANGES_REQUESTED** [review]

File: src/hostapi/jack/pa_jack.c
Comment (Be-ing) [reply src/hostapi/jack/pa_jack.c:470]: This works though

File: src/hostapi/jack/pa_jack.c
Comment (Be-ing) [reply src/hostapi/jack/pa_jack.c]: What we are concerned with here is not the chars + `\0`, it is the size we have malloced. So I think `string_buffer_size` is better than just `string_size`.

File: src/hostapi/jack/pa_jack.c
Comment (Be-ing) [reply src/hostapi/jack/pa_jack.c]: Good idea, this simplifies the code

File: src/hostapi/jack/pa_jack.c
Comment (Be-ing) [reply src/hostapi/jack/pa_jack.c]: yes that works to copy the `\0` terminator

**Comment by Be-ing** [comment]
Thank you for the suggestions @daschuer. That simplified the code.

**Comment by Be-ing** [comment]
@RossBencina @philburk I think this is ready to merge. Please review.

**Review by daschuer — APPROVED** [review]

**Comment by daschuer** [comment]
I am not sure about removing the monitor channels. There are there for a reason and if is not the responsibility of a wrapper library to hide them. I think we need to add a renaming layer that makes them usable with Portaudio. Alternative we may make the ignore list accessible from the outside.

Can you put that in another PR to not slow down the integration process about this?

**Comment by Be-ing** [comment]
I am quite confident about hiding the monitor channels. I can't actually figure out a reason why they exist. It doesn't make any sense to me and I've asked multiple times on the PipeWire issue tracker. Loopback inputs are a hack around the ability to arbitrarily route audio between applications, so they're unnecessary in PipeWire. Exposing them in PortAudio would be confusing because there is no context. In JACK applications like Catia or QJackCtl, the user can see the port names. In PortAudio all we have are devices with a number of channels without any names for the channels.

**Comment by Be-ing** [comment]
Not hiding the monitor channels from would IMO be a significant regression compared to using PortAudio with jackd.

**Comment by Be-ing** [comment]
Not hiding the monitor channels from would IMO be a significant regression compared to using PortAudio with jackd.

**Comment by daschuer** [comment]
I still don't get the point why the application using portaudio should not decide over this. But anyway this is unrelated to the original PR. Please split that out to not delay both.

**Comment by Be-ing** [comment]
The reason is that the PortAudio application *can't* decide about it because PortAudio's API only has concepts of devices with numbers of channels with no additional metadata.

**Comment by Be-ing** [comment]
https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/736

**Comment by daschuer** [comment]
I think a separate monitor device as proposed by @saivert is the way to go.
Let's postpone the issue until upstream has decided.

**Comment by philburk** [comment]
Thanks for this PR.

> I can't actually figure out a reason why they exist.

Loopback channels are not often used.  But they can be handy for doing automated testing.

We are very close to getting 19.7 out. So I am hesitant to include this new code right before a stable snapshot.
Hopefully we will move to a more regular cadence for stable releases.

**Comment by Be-ing** [comment]
PortAudio apps will be broken when Fedora 34 and Ubuntu 21.04 release in a few months without this.

**Comment by Be-ing** [comment]
I can split the last commit to another branch, but the regex escaping is unambiguously something PortAudio needs to do.

**Comment by Be-ing** [comment]
It would be better if PipeWire didn't create the confusing monitor ports by default that probably few users will use but confuse lots of others so PortAudio doesn't have to hack around them.

File: src/hostapi/jack/pa_jack.c
Comment (philburk) [inline src/hostapi/jack/pa_jack.c:543]: jack_client_name_size() include the NUL character so this is OK.

**Review by philburk — APPROVED** [review]

**Review by philburk — COMMENTED** [review]
I changed my approval to a comment that the code looks good. But we should not merge it into 19.7 because it hides the monitor channels. That may be a problem but it is not clear.

**Comment by Be-ing** [comment]
I removed the commit hiding the monitor channels.

**Review by philburk — CHANGES_REQUESTED** [review]
Hmm. It did not remove my approval. I will try clearing that by requesting changes.

**Review by philburk — COMMENTED** [review]
Now setting back to "comment". We may or may not want to hide the monitor channels. That is not clear yet.

**Comment by Be-ing** [comment]
@philburk can you clarify if it is okay with you to merge this for 19.7 without the commit hiding the loopback monitor ports?

**Review by philburk — APPROVED** [review]
Ah! Your commits and my reviews crossed in the mail.
I see you removed the code that hides the channels.
As you say, this code fixes a problem that may break apps in the near future.
I am approving it and we should consider this for 19.7.
Thanks for the contribution.

**Comment by Be-ing** [comment]
The [usability issue with the monitor ports](https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/736) has been solved in PipeWire by presenting them to JACK as separate input devices.

**Comment by RossBencina** [comment]
Looks OK at a glance, but still I need to review it properly. But this is too big a change to risk making days before a stable release.

**Comment by philburk** [comment]
I agree that it is a bit too risky for merging right before a release. Let's address it ASAP after we tag the 19.7 release.

**Comment by Be-ing** [comment]
Now that the usability issue with the loopback monitoring ports has been resolved in PipeWire, this PR only affects a minority of users who have devices that happen to have a regex token in their device name. So it is probably not a big deal to wait on it.

**Comment by Be-ing** [comment]
It seems this is not as esoteric of an issue as I thought. More users [are affected by this](https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/788). Please reconsider merging this for 19.7.

**Comment by dbenjaminmiller** [comment]
Yeah, this issue prevents me from using any PortAudio programs with pw-jack.

**Comment by Holzhaus** [comment]
> Yeah, this issue prevents me from using any PortAudio programs with pw-jack.

Same here.

**Comment by Be-ing** [comment]
I am once again asking you to reconsider merging this for 19.7. Please consider the impact not merging this will have. It would cause user support hassles and reflect badly upon all PortAudio applications on Linux, PipeWire, and Linux distributions. Telling users to compile an unmerged branch of one of an application's dependencies to make the application usable is not a good situation. This may not matter to you personally, but lots of Linux users still have bad memories of the rollout of PulseAudio and are quite concerned that they'll have to experience that again with PipeWire. Please don't contribute to making the rollout of PipeWire rough.

**Comment by Be-ing** [comment]
I would be happy to reach out to the maintainers of the Fedora and Ubuntu portaudio packages and urge them to update to 19.7 for their upcoming releases that will enable PipeWire by default if this gets merged for 19.7.

**Comment by prez** [comment]
I see that this did not make it into 19.7, which is a shame. 
Could you please merge this ASAP? In search for a solution to my audio problems, I switched to pipewire, and this seems to be the final piece preventing it (I use a couple of PortAudio applications) from being a viable alternative.

**Comment by Be-ing** [comment]
@RossBencina can you please review this? It is only +84 −10.

**Comment by rolandlo** [comment]
Having this PR merged would also benefit Xournal++. We depend on Portaudio working fine. More and more users will use PipeWire. It's sad if they get into conflict with apps using Portaudio.

**Comment by Swiftb0y** [comment]
Considering PipeWire is getting increasingly popular (fedora switching to it as a default for example), it would be a shame if this patch won't be included.

**Comment by uklotzde** [comment]
The in-place regex escape code might not be optimal but it looks reasonable and has been confirmed to work. I guess performance is not an issue here and could be improved later.

What matters more is that, as multiple users already confirmed, without this patch PortAudio + Jack would be dysfunctional in all distributions that are switching to PipeWire. It is expected to cause a lot of hassle and support request for applications that rely on PortAudio. We are the ones that users will blame and contact first if things are not working as expected.

**Comment by Holzhaus** [comment]
@RossBencina Please reconsider including this fix in the next stable release. Without this, portaudio apps are unusable with JACK via pipewire on my machine.

File: src/hostapi/jack/pa_jack.c
Comment (Be-ing) [reply src/hostapi/jack/pa_jack.c:480]: Sure, I don't care about the implementation details as long as it works and doesn't crash.

**Comment by Be-ing** [comment]
Thank you for reviewing. I will take care of the review comments tonight. It is unfortunate that something as common as find and replacing in a string is so complicated and error prone in C.

I have just tested installing Mixxx in a fresh Fedora 34 beta virtual machine and finally it Just Works with the PortAudio JACK backend using PipeWire without grabbing exclusive access to the ALSA device or having to configure anything. Of course my RME Babyface Pro does not work in the VM because of this bug so I am eager to get this fix merged.

File: src/hostapi/jack/pa_jack.c
Comment (Be-ing) [reply src/hostapi/jack/pa_jack.c:480]: My original implementation allocated a temporary buffer but I changed it to memmove at @daschuer's suggestion. I don't really care either way.

File: src/hostapi/jack/pa_jack.c
Comment (Be-ing) [reply src/hostapi/jack/pa_jack.c]: Good point. Fixed: ca361337d5ce8df34710fd04ce7662cd3fa7bd2b

File: src/hostapi/jack/pa_jack.c
Comment (Be-ing) [reply src/hostapi/jack/pa_jack.c:480]: I am not sure your proposed method is a better approach. It only works for this one very specific task of escaping regex chars whereas my current implementation can be reused in the future for replacing arbitrary strings in arbitrary strings. Nevertheless I don't want to bikeshed over this. If you insist on changing the approach that you proposed, I will do it.

**Comment by Be-ing** [comment]
@RossBencina The last commit changes to your implementation with some small changes for readability. In my experience, I find it confusing to manipulate a variable with the `++` operator or call a function with side effects in a variable assignment statement, so I moved the `dest++` calls to their own lines. I would prefer to merge everything except the last commit because the `replace_string` function is more flexible and could be used for other situations in the future. But I'm not going to bikeshed over this implementation detail; both implementations work. I tested the current state of this branch with the following devices using the latest PipeWire master branch (db85339f509e8c1bc39bb90fd6a4e6327b257964) and confirm they all work with Mixxx using the PortAudio JACK backend:

```
Babyface Pro (Class Compliant Mode) Pro
Traktor Kontrol S4 MK3 Pro
Built-in Audio Analog Stereo
XONE K2 Analog Surround 4.0
```

File: src/hostapi/jack/pa_jack.c
Comment (daschuer) [inline src/hostapi/jack/pa_jack.c:478]: This is redundant with the else branch

**Review by daschuer — APPROVED** [review]
Only a minor nits. the rest LGTM.

File: src/hostapi/jack/pa_jack.c
Comment (Be-ing) [reply src/hostapi/jack/pa_jack.c]: Good point. I simplified this and force pushed.

**Review by philburk — APPROVED** [review]
Thanks Be-ing for finding and fixing the problem. Also thanks for explaining the impact of not fixing this bug.
Thanks to the reviewers.
And thanks Ross for the simplified regex escape function. It is very clear, which reduces the risk of this late change.

**Comment by Be-ing** [comment]
Great, thank you for merging. It will be easier to get distributions to package this fix now that it has been merged upstream instead of trying to convince them to ship a patch. I have opened a bug report for Fedora to update. @Holzhaus could you notify the Arch packagers to update PortAudio when 19.7 is released? Does anyone want to volunteer to reach out to Debian & Ubuntu?

**Comment by Holzhaus** [comment]
On Arch, it's possible to flag a package as out of date here: https://archlinux.org/packages/extra/x86_64/portaudio/
I can do that when portaudio 19.7 is relased, but in my experience the Arch package maintainers deal with updates in less than a day anyway.

**Comment by Be-ing** [comment]
This fix will be included in Fedora 34 on day 0: https://bugzilla.redhat.com/show_bug.cgi?id=1939749#c4

**Pull request merged**
